// pluggable-cache/lib/index.js
// Generated by CoffeeScript 1.4.0
(function() {
  var Cache, PluggableStore, async, create;

  PluggableStore = require('pluggable-store');

  async = require('async');

  Cache = (function() {

    function Cache(_arg) {
      this.cache = _arg.cache, this.persistence = _arg.persistence, this.lazy = _arg.lazy;
    }

    Cache.prototype.createdStore = function(cb) {
      var isCreated;
      isCreated = function(each, cb) {
        return each.createdStore(cb);
      };
      return async.every([this.cache, this.persistence], isCreated, cb);
    };

    Cache.prototype.createStore = function(cb) {
      var createEach;
      createEach = function(each, cb) {
        return each.createdStore(function(err, created) {
          if (created) {
            return cb(null);
          } else {
            return each.createStore(cb);
          }
        });
      };
      return async.forEach([this.cache, this.persistence], createEach, cb);
    };

    Cache.prototype.removeStore = function(cb) {
      return async.forEach([this.cache, this.persistence], (function(each, cb) {
        return each.removeStore(cb);
      }), cb);
    };

    Cache.prototype.write = function(path, data, cb) {
      if (this.lazy) {
        this.cache.write(path, data, cb);
        return this.persistence.write(path, data, function() {});
      } else {
        return async.forEach([this.cache, this.persistence], (function(each, cb) {
          return each.write(path, data, cb);
        }), cb);
      }
    };

    Cache.prototype.read = function(path, cb) {
      var obj;
      obj = this;
      return this.cache.read(path, function(err, res) {
        if (err) {
          return obj.persistence.read(path, function(err, res) {
            obj.cache.write(path, res, function() {});
            return cb(null, res);
          });
        } else {
          return cb(null, res);
        }
      });
    };

    Cache.prototype.remove = function(path, cb) {
      if (this.lazy) {
        this.cache.remove(path, cb);
        return this.persistence.remove(path, function() {});
      } else {
        return async.forEach([this.cache, this.persistence], (function(each, cb) {
          return each.remove(path, cb);
        }), cb);
      }
    };

    Cache.prototype.keys = function(cb) {
      var obj;
      obj = this;
      return this.cache.keys(function(err, res) {
        if (err) {
          return obj.persistence.keys(cb);
        } else {
          return cb(null, res);
        }
      });
    };

    return Cache;

  })();

  create = function(opts) {
    return new PluggableStore({
      adapter: new Cache(opts)
    });
  };

  create.adapter = Cache;

  module.exports = create;

}).call(this);
