// pluggable-store/lib/filesystem.js
// Generated by CoffeeScript 1.4.0
(function() {
  var FileSystem, exec, fs, removeDir, resolvePath;

  fs = require('fs');

  resolvePath = (require('path')).resolve;

  exec = require('child_process').exec;

  removeDir = function(dir, cb) {
    return exec('rm -r -f ' + dir, cb);
  };

  FileSystem = (function() {

    function FileSystem(rootPath) {
      this.rootPath = rootPath;
    }

    FileSystem.prototype.exists = function(cb) {
      return fs.exists(this.rootPath, function(exists) {
        return cb(null, exists);
      });
    };

    FileSystem.prototype.create = function(cb) {
      return fs.mkdir(this.rootPath, cb);
    };

    FileSystem.prototype.destroy = function(cb) {
      return removeDir(this.rootPath, cb);
    };

    FileSystem.prototype.write = function(path, data, cb) {
      return fs.writeFile(this.path(path), data, 'utf8', cb);
    };

    FileSystem.prototype.read = function(path, cb) {
      return fs.readFile(this.path(path), 'utf8', cb);
    };

    FileSystem.prototype.remove = function(path, cb) {
      return fs.unlink(this.path(path), cb);
    };

    FileSystem.prototype.path = function(fileName) {
      return resolvePath(this.rootPath, fileName);
    };

    FileSystem.prototype.keys = function(cb) {
      return fs.readdir(this.rootPath, cb);
    };

    return FileSystem;

  })();

  module.exports = FileSystem;

}).call(this);
